primitives/staking/src/lib.rs

```diff
@@ -135,6 +135,32 @@ pub trait StakingInterface {
 		num_slashing_spans: u32,
 	) -> Result<bool, DispatchError>;
 
+	/// Get StakingLedger by controller account.
+	fn get_ledger(ctrl: &Self::AccountId) -> StakingLedger<Self>;
```

Comment:
@kianenigma We might want to get rid of those, seems redundant now.:
```
	fn active_stake(controller: &Self::AccountId) -> Option<Self::Balance> {
		Self::ledger(controller).map(|l| l.active)
	}

	fn total_stake(controller: &Self::AccountId) -> Option<Self::Balance> {
		Self::ledger(controller).map(|l| l.total)
	}
```

Reply:
Now I see why the old code was not returning `StakingLedger`. We cannot use `StakingLedger` in `pallet-nomination-pools` or `pallet-fast-unstake` because it is defined in pallet-stakin.

Putting it in a shared crate is also too much of a stretch in my opinion, because these types are arguably an implementation detail.

I think we should keep these, not have `get_ledger` and `get_exposure` and instead create other API function that express a condition, a state or a behavior rather than the explicit type.

Reply:
Why don't we introduce a special type that will take values out of StakingLedger? Like type `Ledger`. It could even be an alias for now and then change if needed. 

This makes very little sense to keep as is, because:
```
			let ledger =
				pallet_staking::Ledger::<T>::get(&ctrl).ok_or(Error::<T>::NotController)?;
			ensure!(!Queue::<T>::contains_key(&ledger.stash), Error::<T>::AlreadyQueued);
			ensure!(
				Head::<T>::get().map_or(true, |UnstakeRequest { stash, .. }| stash != ledger.stash),
				Error::<T>::AlreadyHead
			);
			// second part of the && is defensive.
			ensure!(
				ledger.active == ledger.total && ledger.unlocking.is_empty(),
				Error::<T>::NotFullyBonded
			);
```

We'd need to call three different methods that operate on exactly the same storage item. Actually no, 4!
At the very least that's not very dry and also a bit of overhead.

Reply:
my answer to this and many other aspects can be found in https://github.com/paritytech/substrate/compare/ru/feature/fast-unstake-istaking...kiz-ru/feature/fast-unstake-istaking